### DISCUSSION TITLE
Not working on K3

### DISCUSSION THREAD

**Discussion by healthyfat** [koreader/koreader#11246] [Q&A]
Am I doing something wrong? It doesn't start on my Kindle Keyboard. Log file shows it is missing some files:

It's a scroll... It's a codex... It's KOReader!

 [*] Current time: 12/20/23-20:39:22
 [*] Version: v2023.10

ffi.load: libs/libutf8proc.so.2
ffi.load: blitbuffer
12/20/23-20:39:23 WARN  could not load liblipclua: module 'liblipclua' not found:
	no field package.preload['liblipclua']
	no file 'common/liblipclua.lua'
	no file 'rocks/share/lua/5.1/liblipclua.lua'
	no file 'frontend/liblipclua.lua'
	no file './liblipclua.lua'
	no file '/usr/local/share/luajit-2.1/liblipclua.lua'
	no file '/usr/local/share/lua/5.1/liblipclua.lua'
	no file '/usr/local/share/lua/5.1/liblipclua/init.lua'
	no file 'common/liblipclua.so'
	no file 'common/liblipclua.dll'
	no file '/usr/lib/lua/liblipclua.so'
	no file 'rocks/lib/lua/5.1/liblipclua.so'
	no file './liblipclua.so'
	no file '/usr/local/lib/lua/5.1/liblipclua.so'
	no file '/usr/local/lib/lua/5.1/loadall.so' 
12/20/23-20:39:23 WARN  could not load liblibclua 
[ko-input] Forked off fake event generator (pid: 1243)
12/20/23-20:39:23 INFO  initializing for device Kindle3 
12/20/23-20:39:23 INFO  framebuffer resolution: {
  h = 800,
  w = 600
} --[[table: 0x405a33f0]] 
ffi.load: libs/libmupdf.so
ffi.load: libs/libwrap-mupdf.so
ffi.load: sqlite3
12/20/23-20:39:24 INFO  Performing one-time migration for 20191129 
12/20/23-20:39:24 INFO  Performing one-time migration for 20200421 
12/20/23-20:39:24 WARN  os.remove: ./cache/fontlist/fontinfo.dat: No such file or directory 
ffi.load: libs/libfreetype.so.6
ffi.load: libs/libharfbuzz.so.0
ffi.load: libs/libzstd.so.1
12/20/23-20:39:25 INFO  ./cache/fontlist/fontinfo.dat fopen: No such file or directory -> initializing it 
12/20/23-20:39:28 INFO  Performing one-time migration for 20210306 (1/2) 
12/20/23-20:39:28 INFO  Performing one-time migration for 20210306 (2/2) 
12/20/23-20:39:28 INFO  Performing one-time migration for 20210330 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210404 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210409 
12/20/23-20:39:29 WARN  os.rename: ./cache/fontinfo.dat: No such file or directory 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210412 
12/20/23-20:39:29 WARN  os.rename: ./cache/calibre-libraries.lua: No such file or directory 
12/20/23-20:39:29 WARN  os.rename: ./cache/calibre-books.dat: No such file or directory 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210414 
12/20/23-20:39:29 WARN  os.remove: ./cache/calibre/books.dat: No such file or directory 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210503 & 20210508 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210622 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210521 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210531 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210629 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210715 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210720 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210831 
12/20/23-20:39:29 INFO  Performing one-time migration for 20210925 
12/20/23-20:39:29 INFO  Performing one-time migration for 20220116 
12/20/23-20:39:29 INFO  Performing one-time migration for 20220205 
12/20/23-20:39:29 INFO  Performing one-time migration for 20220819 
12/20/23-20:39:29 INFO  Performing one-time migration for 20220914 
12/20/23-20:39:29 INFO  Performing one-time migration for 20220930 
12/20/23-20:39:29 WARN  loadfile: cannot open ./defaults.persistent.lua: No such file or directory 
12/20/23-20:39:29 WARN  os.rename: ./defaults.persistent.lua: No such file or directory 
12/20/23-20:39:29 INFO  Performing one-time migration for 20221027 
12/20/23-20:39:29 INFO  Performing one-time migration for 20230531 
12/20/23-20:39:29 INFO  Performing one-time migration for 20230703 
12/20/23-20:39:29 INFO  Performing one-time migration for 20230707 
12/20/23-20:39:29 INFO  Performing one-time migration for 20230710 
12/20/23-20:39:29 INFO  Performing one-time migration for 20230731 
12/20/23-20:39:29 INFO  Performing one-time migration for 20230802 
12/20/23-20:39:30 INFO  Performing one-time migration for 20230901 
12/20/23-20:39:32 INFO  opening file ./help/quickstart-en-v2023.10.html 
CRE: font PMN Caecilia LT regular: updated weight from 500 to 400
12/20/23-20:39:33 INFO  Inhibiting user input 
12/20/23-20:39:33 INFO  Loading plugins from directory: plugins 
sh: command: not found
sh: command: not found

**Comment by NiLuJe** [comment]
Nope, that's a perfectly normal log (the lua lipc bindings are only found on FW 5.x).

**Comment by NiLuJe** [comment]
(And the `command` failures are probably from the timesync plugin, so, again, harmless).

**Reply by healthyfat** [reply]
Good to know. Any idea why it is not starting then?

**Reply by NiLuJe** [reply]
Is *is* starting (and successfully at that), so you'll have to be more specific.

**Reply by healthyfat** [reply]
It seems like it stuck when starting. It shows message:  opening file ./help/quickstart-en-v2023.10.html and nothing happens even after waiting over 5 mins. I guess it shouldn't take that long to start. 

Those are final logs before I restart device: 

12/20/23-20:39:32 INFO opening file ./help/quickstart-en-v2023.10.html
CRE: font PMN Caecilia LT regular: updated weight from 500 to 400
12/20/23-20:39:33 INFO Inhibiting user input
12/20/23-20:39:33 INFO Loading plugins from directory: plugins
sh: command: not found
sh: command: not found

**Reply by Frenzie** [reply]
I'm slightly surprised by how it takes 9 seconds before it gets around to opening the quickstart guide, even if a first-time start would take a little longer. I suppose that must be normal on what's pretty much the oldest/slowest type of supported device. But even given that I doubt it should take much more than let's say 20 seconds total, yeah.

**Reply by NiLuJe** [reply]
How are you launching KOReader?

**Reply by healthyfat** [reply]
With KUAL.

**Reply by NiLuJe** [reply]
Welp, more on this if my K3 ever wakes up from its sleep.

**Reply by NiLuJe** [reply]
Yup, it's hanging in a `FUTEX_WAIT` somewhere inside CRe, during `DEBUG CreDocument: rendering document...`

Going to check if a newer GCC version doesn't magically fix it...

Funnily enough, it appears to only hang when attempting to show the QuickStart Guide... If I bypass that, everything looks dandy...

**Reply by NiLuJe** [reply]
Assuming it's a clean install, you can bypass the quickstart with a custom config (as `koreader/settings.reader.lua`).

```lua
-- ./settings.reader.lua
return {
    ["quickstart_shown_version"] = 202310000066,
}
```

**Reply by NiLuJe** [reply]
> Funnily enough, it appears to only hang when attempting to show the QuickStart Guide...

Or, more generically, if `start_with` is set to `last` *and* the last document is opened by CRe (e.g., a PDF works just fine). Huh.

**Reply by NiLuJe** [reply]
> Going to check if a newer GCC version doesn't magically fix it...

Well, that went nowhere, as newer STL libraries pull in glibc 2.6 symbols...

**Reply by mergen3107** [reply]
Quick Start is an html file? Maybe that's the problem? Style tweaks?

**Reply by healthyfat** [reply]
@NiLuJe Are you suggesting to create file settings.reader.lua in koreader folder to bypass opening file ./help/quickstart-en-v2023.10.html ?

**Reply by Frenzie** [reply]
Yup. :)

**Reply by healthyfat** [reply]
Unfortunately it didn't work. I created 

settings.reader.lua 

file containing provided script and placed it in koreader main folder. Unless I did something wrong as I am obviously a novice :) 

Script:

-- ./settings.reader.lua
return {
    ["quickstart_shown_version"] = 202310000066,
}

**Reply by Frenzie** [reply]
By didn't work you mean it still shows it's trying to load that file in the log or that it gets stuck on something else?

**Reply by NiLuJe** [reply]
Then you bungled up the settings file ;).

**Reply by healthyfat** [reply]
Sorry. It worked. I used different script by mistake. Thank you.

**Reply by NiLuJe** [reply]
Deadlocks during a local static initialization. (Reliably, at the same place).

It's either an actual code issue (usually involving recursion), an old STL issue (but still present in GCC 9.5), or an issue on the hilariously old host glibc (2.5) or kernel (2.6.26), as it is, after all, hanging on a syscall (futex).

```
#0  0x4019a110 in syscall () from /lib/libc.so.6
No symbol table info available.
#1  0x410aa014 in __cxa_guard_acquire () from /mnt/us/koreader/./libs/libstdc++.so.6
No symbol table info available.
#2  0x41b8bd68 in lunasvg::elementid (name=...) at /var/tmp/niluje/koreader/base/thirdparty/lunasvg/build/arm-kindle-linux-gnueabi-debug/lunasvg-prefix/src/lunasvg/source/parser.cpp:985
        elementmap = {_M_t = {
            _M_impl = {<std::allocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, lunasvg::ElementID> > >> = {<__gnu_cxx::new_allocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, lunasvg::ElementID> > >> = {<No data fields>}, <No data fields>}, <std::_Rb_tree_key_compare<std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >> = {
                _M_key_compare = {<std::binary_function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool>> = {<No data fields>}, <No data fields>}}, <std::_Rb_tree_header> = {_M_header = {_M_color = std::_S_red, _M_parent = 0x0, 
                  _M_left = 0x0, _M_right = 0x0}, _M_node_count = 0}, <No data fields>}}}
        it = <optimized out>
#3  0x41b8fba8 in lunasvg::TreeBuilder::parse (this=this@entry=0xbe8218c0, data=data@entry=0x313600 "<svg style=\"font-size: 22;font-family: Noto Serif;font-weight: 400;color: #000000;\" lang=\"en\" xmlns=\"http://www.w3.org/2000/svg\" xlink=\"http://www.w3.org/1999/xlink\" version=\"1.1\" width=\"100%\" height="..., size=size@entry=322)
    at /var/tmp/niluje/koreader/base/thirdparty/lunasvg/build/arm-kindle-linux-gnueabi-debug/lunasvg-prefix/src/lunasvg/source/parser.cpp:1914
```

(Rest of the trace omitted, as it's long and mostly uninteresting, and gdb will OOM the system trying to get to the bottom of it ;D).

**Reply by NiLuJe** [reply]
https://github.com/koreader/koreader-base/pull/1717/commits/adb63980525126dc3aaace0f98a7c72dcc91251d will paper over it.

**Reply by healthyfat** [reply]
I know it is not directly connected to KOreader, but I can't find answer elsewhere. 

I have problem connecting to K3 using WinSCP. I have USBnet enabled and I followed all the instruction, but I am not able to connect. 

WinSCP can't find Kindle even thought it is present in Network Connections as Kindle USB RNDIS Device and I am getting Network error: Connection timed out. Ping 192.168.15.244 doesn't work as well. 

I have Windows 10 and I use below settings:

Host IP:192.168.15.244
User: root

Kindle IP:  192.168.15.201
subnet mask: 255.255.255.0

**Reply by NiLuJe** [reply]
If you're confident about the networking aspect being right, try from a real OS with known good drivers for CDC ethernet (e.g., Linux or macOS).

Your win 10 might have inherited a crap RNDIS driver from... something along the way.

(IIRC, win11 is... slightly less stupid).

**Reply by healthyfat** [reply]
I get it working. I was confident about IP address cuz all the instructions suggest same one, but they are wrong. At least for K3. It should be 192.168.2.1 and the host 192.168.2.2. Thanks

**Reply by NiLuJe** [reply]
Oh, right, the .15 subnet is the default for diags and FW 5.x, older devices indeed default to .2
